### PR TITLE
Support for TypedArray interop between WASM and JS

### PR DESCRIPTION
**Pull request by kyr0**
As discussed in https://github.com/AssemblyScript/assemblyscript/issues/263 and https://github.com/AssemblyScript/assemblyscript/issues/240, I added support for memory mapped interop of all standard TypedArray subtypes. I added tests for every subtype to ensure a safe operation. Type checks ensure the safe construction of TypedArray subtypes in the loader lib.

### CODE REVIEW

File: lib/loader/index.js [hunk 1]
```diff
@@ -71,6 +71,53 @@ function instantiate(module, imports) {
...
 
+  /** Allocates a new TypedArray in the module's memory and returns it's pointer and TypedArray object instance reference */
+  function newTypedArray(length, typedArrayCtor) {
+
+    let bytesToAllocate = 0,
+        ptr,
+        view;
+
+    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays
+    if (typedArrayCtor === Int8Array || typedArrayCtor === Uint8Array || typedArrayCtor === Uint8ClampedArray) {
```
Comment (MaxGraey) [inline lib/loader/index.js]: Just use [<TypedArray>.BYTES_PER_ELEMENT](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/BYTES_PER_ELEMENT)

**Comment by MaxGraey** [comment]
```js
function newBuffer(view, length, unsafe) {
  return view.constructor === Function
      ? newEmptyBuffer(view, length, unsafe)
      : newCopiedBuffer(view, length);
}
```
?

**Comment by dcodeIO** [comment]
That's just so you can do

```js
newBuffer(new Float32Array([1.1, 1,2, ...]);
```

or

```
newBuffer(Float32Array, 10);
```

**Comment by MaxGraey** [comment]
I see. I just add `unsafe` param, so we can do like this `newBuffer(Float32Array, 10, false);` without clear buffer area

**Comment by MaxGraey** [comment]
And when we need read buffer after procces in js side we just create this:
```js
const ptr = newBuffer(Float32Array, 10);
instance.exports.process(ptr, 10);
const bufF32 = new Float32Array(instance.exports.memory.buffer, ptr, 10);
// read processed result from bufF32
```
Right?

**Comment by dcodeIO** [comment]
Either that or return the respective view as well as proposed in the PR.

**Comment by MaxGraey** [comment]
```js
function getBuffer(ctor, ptr, length) {
   checkMem();
   return new ctor(mem, ptr, length);
}

getBuffer(Float32Array, ptr, 10);
```

**Comment by MaxGraey** [comment]
I prefer create Typed Array lazily and only when we really need read back from memory

**Comment by MaxGraey** [comment]
And will be great ability create ArrayBuffer from Pointer on AssemblyScript but this require extra space (for 4-byte length field) during alloc inside newArray.

**Comment by dcodeIO** [comment]
For reference, for a real typed array on the AS side we'd need to do something like this (didn't test):

```js
  function computeBufferSize(byteLength) {
    const HEADER_SIZE = 8;
    return 1 << (32 - Math.clz32(byteLength + HEADER_SIZE - 1));
  }

  /** Creates a new typed array in the module's memory and returns its pointer. */
  function newArray(view, length, unsafe) {
    var ctor = view.constructor;
    if (ctor === Function) { // TypedArray constructor
      ctor = view;
      view = null;
    } else { // TypedArray instance
      if (length === undefined) length = view.length;
    }
    var elementSize = ctor.BYTES_PER_ELEMENT;
    if (!elementSize) throw Error("not a typed array");
    var byteLength = elementSize * length;
    var ptr = exports["memory.allocate"](12); // TypedArray header
    var buf = exports["memory.allocate"](computeBufferSize(byteLength)); // ArrayBuffer
    checkMem();
    U32[ ptr      >>> 2] = buf;        // .buffer
    U32[(ptr + 4) >>> 2] = 0;          // .byteOffset
    U32[(ptr + 8) >>> 2] = byteLength; // .byteLength
    U32[ buf      >>> 2] = byteLength; // .byteLength
    U32[(buf + 4) >>> 2] = 0;          // 0
    if (view) {
      new ctor(mem, buf + 8, length).set(view);
    } else if (!unsafe) {
      exports["memory.fill"](buf + 8, 0, byteLength);
    }
    return ptr;
  }

  /** Gets a view on a typed array in the module's memory by its pointer. */
  function getArray(ctor, ptr) {
    var elementSize = ctor.BYTES_PER_ELEMENT;
    if (!elementSize) throw Error("not a typed array");
    checkMem();
    var buf        = U32[ ptr      >>> 2];
    var byteOffset = U32[(ptr + 4) >>> 2];
    var byteLength = U32[(ptr + 8) >>> 2];
    return new ctor(mem, buf + 8 + byteOffset, (byteLength - byteOffset) / elementSize);
  }
```

**Comment by MaxGraey** [comment]
```js
...
U32[(ptr + 8) >>> 2] = byteLength; // .byteLength
U32[ buf      >>> 2] = byteLength; // .byteLength
...
```
Isn't this will be:
```js
U32[(ptr + 8) >>> 2] = length; // .length
U32[ buf      >>> 2] = byteLength; // .byteLength
```
?

**Comment by dcodeIO** [comment]
Layout of a typed array differs a bit from a normal array. A typed array computes the length property from its byteLength and type.

**Comment by MaxGraey** [comment]
Got it! I thought this usual Array

**Comment by MaxGraey** [comment]
Hmm, btw `length` from Array. Could we drop this and calculate `length` from ArrayBuffer's `byteLength` same as we calculate length for TypedArray?
```ts
@inline
get length(): i32 {
    return this.buffer_.byteLength >> alignof<T>();
}
```

**Comment by dcodeIO** [comment]
We can't drop it there because a user can modify `.length`. For instance, the backing buffer might have space for 100 elements, but the length is still 0 until values are pushed to it, which the length property keeps track of.

**Comment by MaxGraey** [comment]
Got it, so it actually works as capacity

**Comment by dcodeIO** [comment]
As this seems to lead to some confusion, I decided to go ahead with an implementation that creates a proper typed array for use on the AS side [here](https://github.com/AssemblyScript/assemblyscript/commit/9c770d801eaff438c2d940b22cee651234597fd0).

So, this should "just work":

```ts
// JS side
var ptr = module.newArray(new Int32Array([1, 2, 3]));
module.processArray(ptr);
console.log(module.getArray(ptr));

// AS side
export function processArray(arr: Int32Array): void {
  ...
}
```

**Comment by dcodeIO** [comment]
> I find it counter-intuitive that getArray() returns a TypedArray subclass instance

Hmm, might be, yeah. My thought was that all we can do currently is typed arrays anways, but if you think it helps we can rename it to `getTypedArray` etc.

**Comment by jbousquie** [comment]
I know that naming things is always a great problem, so just a semantic suggestion : wouldn't be more pertinent to call it `newASArray()` or `ASArray()` in order to not get confused with standard JS arrays, knowing its name is written in the code JS side anyway ? 
It's a new Array type and it's pure AS, soI would have called it an ASArray in first intention if I didn't know its name.
Only my (newbee) opinion.

**Comment by dcodeIO** [comment]
> Did you try my initial PR? It returns the reference to the constructor function of the TypedArray together with it's instance and pointer so you don't need to put it again on JS side.

Yeah, I was just concerned about the case where a typed array is returned from AS. That'd be just a number and it seemed more consistent this way.

**Comment by kyr0** [comment]
> Yeah, I was just concerned about the case where a typed array is returned from AS. That'd be just a number and it seemed more consistent this way.

Oh well, on AS side my PR only differs in passing two parameters because I didn't knew I could just use the TypedArray counterpart as an argument in AS. I also like your proposal much more for the AS side. This is clearly more elegant.

I think the proposed API of my PR (JS side) and yours (JS and AS side) could go hand in hand... 

Bringing @jbousquie remark into play... as it is a new use case... like "sharing a TypedArray" between JS and AS we could maybe name the methods for what they do context-wise... 

`newArray()` could be `newSharedTypedArray` and return an object like `ShareDescriptor`  containing the actual TypedArray `instance`, `ptr` and the `constructor reference` to the constructor of the TypedArray subtype..., the call to AS side could still do the same as it does now and referencing to the constructor again wouldn't be necessary.

It would become:

    // shared:
    // shared.ptr -> number
    // shared.array -> actual instance of Int32Array
    // shared.type -> constructor reference (Int32Array)
    var shared = module.newSharedTypedArray(new Int32Array([1, 2, 3]));
    module.processArray(shared.ptr);
    console.log(shared.array); // no need for getArray() anymore
    module.freeArray(shared.type, shared.ptr); // once not used anymore

    // AS side
    export function processArray(arr: Int32Array): void {
        ...
    }

But it's up to you. It's working as it is now and from my point of view it's already doing it's job...

**Comment by MaxGraey** [comment]
> newArray() could be newSharedArray
This could be confused because JS have [`SharedArrayBuffer`](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer) which using for threads

**Comment by jbousquie** [comment]
`newASSharedArray()` ?

**Comment by dcodeIO** [comment]
Regarding naming: I think `newArray` is fine as the typed nature of such an array is (currently) implicit anyway. Once GC/reference-types lands, normal arrays won't have to go through a helper anymore anyway I think.

Hence, closing for now, but feel free to discuss :)

**Pull request closed without merging**